### PR TITLE
[Clang][Driver][LTO] Fix empty stats filename when in LTO mode

### PR DESCRIPTION
**Pull request by mshockwave**
Previously, if a linker flag (i.e. -Wl) is presented before any input filenames, Gnu driver would use the InputInfo object of that flag to generate stats filename for LTO backend, causing an empty filename. This patch fixes such issue.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Gnu.cpp [hunk 1]
```diff
@@ -535,7 +535,15 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
 
   if (D.isUsingLTO()) {
     assert(!Inputs.empty() && "Must have at least one input.");
-    addLTOOptions(ToolChain, Args, CmdArgs, Output, Inputs[0],
+    // Find the first filename InputInfo object.
+    auto Input = llvm::find_if(
+        Inputs, [](const InputInfo &II) -> bool { return II.isFilename(); });
+    if (Input == Inputs.end())
+      // For a very rare case, all of the inputs to the linker are
+      // flags. If that happens, just use the first InputInfo.
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/Gnu.cpp:543]: In Clang, flags only refer to boolean options (i.e. not `-fxxx=yy`). It seems that the canonical term here is `InputArg`.

File: clang/test/Driver/save-stats.c [hunk 2]
```diff
@@ -20,6 +20,8 @@
 // CHECK-INVALID: invalid value 'bla' in '-save-stats=bla'
 
 // RUN: %clang -target x86_64-linux-unknown -save-stats -flto -o obj/dir/save-stats.exe %s -### 2>&1 | FileCheck %s -check-prefix=CHECK-LTO
+// Previously `-plugin-opt=stats-file` would use empty filename if a linker flag (i.e. -Wl) is presented before any input filename.
+// RUN: %clang -target x86_64-linux-unknown -save-stats -flto -o obj/dir/save-stats.exe -Wl,-plugin-opt=-dummy %s -### 2>&1 | FileCheck %s -check-prefix=CHECK-LTO
```
Comment (MaskRay) [inline clang/test/Driver/save-stats.c:24]: Use `--target=` for new tests. `-target ` has been deprecated since Clang 3.4.

**Review by MaskRay — APPROVED** [review]

File: clang/test/Driver/save-stats.c (hunk 2 above)
Comment (mshockwave) [reply clang/test/Driver/save-stats.c]: Done

File: clang/lib/Driver/ToolChains/Gnu.cpp (hunk 1 above)
Comment (mshockwave) [reply clang/lib/Driver/ToolChains/Gnu.cpp]: Done

**Pull request merged**